### ISSUE TITLE
Maybe break save compatibility sometime this decade

### ISSUE THREAD

**Issue by emilazy** [crawl/crawl#1364]
Save compatibility hasn't been broken [since 2012](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/docs/develop/history/major-tag-history.md); there are probably 8 year olds playing Crawl who have never seen a compatibility break. Currently, just running the `tag-major-upgrade` script removes ~10k lines from the source code:

```
emily@renko ~/s/c/c/source> util/tag-major-upgrade -t 35
…
emily@renko ~/s/c/c/source> git diff --unified=0 --minimal | wc -l
13268
```

Most of this code is simple and benign, but there's a huge amount of it, in pretty much every major file in the codebase, and there's definitely [gnarlier](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/files.cc#L1261-L1285) [stuff](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/files.cc#L1893-L1911) [lurking](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L1045-L1059) [if](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L3386-L3404) [you](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L3872-L3903) [go](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L3958-L3967) [looking](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L4335-L4361) [for](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L6211-L6222) [it](https://github.com/crawl/crawl/blob/19abc6cbe401640f82ef30f66498c605bcd7f2b7/crawl-ref/source/tags.cc#L6449-L6604). The number of forgotten relics still kept around in the source tree for save compatibility is astounding; the era of `TAG_MAJOR_VERSION == 34` is the era of Dwarven Hall, of Lava Orcs and Djinni, of orc slaying weapons and bottled efreet; much of the Pakellas code is still lying around for transferred games, even though it was disabled [back in 2016](https://github.com/crawl/crawl/commit/e213fbe9c8345be16549d0715d7deb1c3d1282dc) and has been marked for removal [since the year after](https://github.com/crawl/crawl/commit/9a9e23e64ec5408265be811f5a7b7c896141daf7).

All this added complexity and legacy cruft might not be so bad if players derived a corresponding benefit from it, but in practice I suspect most games are played and concluded within the course of one release. Online players have no way to transfer games between stable versions at all, and anyone who has transferred a trunk save across years of versions knows that the result is frequently a tangle of broken removed items and spells, edge-cases, and sometimes crash bugs. Given that the servers already handle the case where saves can't be transferred across trunk versions, and a player whose save is broken by a botched upgrade has less recourse than one stuck on an old version, it's not clear to me that rigidly keeping save compatibility over the course of years is the best option.

Of course, there's no point in breaking save compatibility for no reason whatsoever, and a return to the frequent breaks of the 0.[678]a era documented in the history I linked would be a regression, but I think it's at least worth strongly considering doing so the next time anything that would require complex save compatibility logic is implemented. (A modest proposal™ would be to not bother keeping save compatibility between stable releases, but I realize that one might be too radical!)

Apologies for the rant; I've been somewhat unconvinced of the benefits of Crawl's obsessive save compatibility policy for years now, but couldn't resist throwing in my 2¢ after seeing just how much code it really is at this point…

**Comment by alexjurkiewicz** [comment]
This would make life much easier for contributors too. It took me a long time to understand TAG_MAJOR_VERSION, especially wrt enums with duplicate entries.

I spent time a few months ago making the codebase compile cleanly with TAG_MAJOR_VERSION 35, see comments in https://github.com/crawl/crawl/pull/955

**Comment by gammafunk** [comment]
If by "this" you mean the suggestion to not keep save compatibility between stable releases, that still requires retaining TAG_MAJOR_VERSION and all the duplication of enums entailed. Removing this logic entirely would mean you could break save compatibility with any arbitrary trunk commit, which is a regression as Emily says. All breaking save compat with each major release means is that we'd do the bump with every release.

**Comment by alexjurkiewicz** [comment]
Yes, I understand the tag framework would remain. If the major version is bumped, a lot of existing compatibility code will be deleted as emily says. Fewer changes will touch areas that have "legacy code".

**Comment by rawlins** [comment]
FWIW I think most active devteam members are broadly in agreement with this, we discuss this at least once every few months. But actually doing it is a big project with a ton of logistics, and no one has been that excited to take it on. It also probably needs to be synced to the timing of the release cycle.  I'm not entirely confident that all servers would uniformly handle this well, since there does seem to be code that would handle this for dgamelaunch-config servers but it's presumably not very well tested, so that's another set of logistics.

Setting all of this discussion aside, one reason we discuss this from time to time is that minor version is stored as a byte and it is running out of space, so there is a ticking clock.

**Comment by emilazy** [comment]
> I spent time a few months ago making the codebase compile cleanly with TAG_MAJOR_VERSION 35, see comments in #955

A good way to ensure that this doesn't regress in this future might be to have CI automatically build and test with a bumped TAG_MAJOR_VERSION; the friction of fixing all the miscellaneous broken stuff undoubtedly contributes to the reluctance to break compatibility.

**Comment by alexjurkiewicz** [comment]
I've started a pull request to add CI support. There is at least one existing error with incorrect TAG_MAJOR_VERSION'ing (related to SPWPN_EXPLODING), and probably more. The goal is to fix all existing issues in that branch.

**Comment by aidanholm** [comment]
I think a better alternative is to remove as many sources of incompatibility a major bump would cause as possible. A large part of this is due to using enum values in saves, which prevents reordering them in headerfiles. An alternative is to replace enum values in saves with incremental ids, and add a mapping of ids to enum names (strings) to the save file. The crawl binary can convert from enum name to enum value.

Benefits:
 - we could reorder enum values freely and remove all the ifdefs
 - incremental conversion is possible (an enum at a time)
 - probably no need to sync with releases or do any special handling with servers
 - this is necessary for mod support anyway
 - it should be a rather straight-forward implementation / conversion
 - should be entirely user-transparent
 - the crawl binary can create an id-to-enum table once for fast loads if required

Shortcomings:
 - slightly larger save files
 - imperceptibly slower game saves and loads

Thoughts?

**Comment by lembark** [comment]
If binary compatibility gets blown once every decade that puts crawl in rather good company for stability.

**Comment by aidanholm** [comment]
There's been a bit of discussion on `##crawl-dev`, with three possible ways we could get rid of this old save compat stuff:
 - rolling support: incrementally remove support for old minors at roughly the same rate we add new ones. See PR #1375. Major stays at 34.
 - windowed support: same as rolling, but minor removals are grouped in large chunks with an accompanying major update. We support some number of previous majors (e.g. M-2, M-1, and M) in a binary with major M.
 - break compat, then switch to one of the above: this lets us immediately remove all ifdef'd code. Much could still be removed after removing old minors, but it's all conditionally compiled based on the major, so some git scripting would be needed to find bits that can be removed.
